torchtune/datasets/_concat.py

```diff
@@ -0,0 +1,134 @@
+from torchtune.modules.tokenizers import Tokenizer
+from tqdm import tqdm
+
+
+class ConcatDataset(Dataset):
+    """
+    Class that enables continued pretraining by packing a single column of text data into equally sized samples.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    Streaming datasets are supported. To stream a dataset, pass `'streaming': True` in the `load_dataset_kwargs`.
+    If streaming, the total number of samples will be unknown until the end of the dataset, and the progress bar
+    will be indeterminate.
+
+    The general flow is:
+
+    On initialization:
+    optionally shuffle sample -> load sample -> tokenize and add to buffer ->
+        when buffer is long enough, add to self.samples.
+
+    During training:
+    return self.samples[idx] as input and label.
+
+    Args:
+        tokenizer (Tokenizer): Tokenizer used to encode data. Tokenize must implement an `encode` and `decode` method.
+        source (str): path string of dataset, anything supported by Hugging Face's `load_dataset`,
+            or if `is_local` is true, a local path to load with `load_from_disk`.
+        text_column (str): column name of the dataset to pack into samples.
+        max_seq_len (int): Maximum number of tokens in the returned input and label token id lists.
+            Default is None, disabling truncation. We recommend setting this to the highest you can fit in memory
+            and is supported by the model. For example, llama2-7B supports up to 4096 for sequence length.
+        max_rows (int): maximum number of samples to pack. Default is None, which will pack as many samples as possible.
+        is_local (bool): whether the source is a local path. Default is False.
+        shuffle_before_packing (bool): whether to shuffle the dataset before packing. Default is False.
+        seed (int): seed for shuffling. Default is 29.
+        train_split_name (str): name of the split to use. Default is None, which will use the full dataset.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to `load_dataset`.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        source: str,
+        text_column: str,
+        max_seq_len: int,
+        max_rows: int = None,
+        is_local: bool = False,
+        shuffle_before_packing: bool = False,
+        seed: int = 29,
+        train_split_name: str = None,
+        **load_dataset_kwargs: Dict[str, Any],
+    ):
+
+        assert tokenizer is not None, "tokenizer must be provided"
+        assert source is not None, "source must be provided"
+        assert text_column is not None, "text_column must be provided"
+        assert max_seq_len is not None, "max_seq_len must be provided"
+
+        if is_local:
+            dataset = load_from_disk(source, **load_dataset_kwargs)
+        else:
+            dataset = load_dataset(source, **load_dataset_kwargs)
```

Comment:
can we just use `load_dataset`'s ability to load from a local file, or is there some missing functionality?